src/python/pants/backend/python/util_rules/interpreter_constraints.py

```diff
@@ -246,6 +247,208 @@ def requires_python38_or_newer(self, interpreter_universe: Iterable[str]) -> boo
+
+        assert valid_py3_patches
+        all_valid_py3_versions = list(valid_py3_patches)
+        first_py3_version = all_valid_py3_versions[0]
+        last_py3_version = all_valid_py3_versions[-1]
+        only_one_py3_version = first_py3_version == last_py3_version
+
+        first_py3_version_patches = valid_py3_patches[first_py3_version]
+        last_py3_version_patches = valid_py3_patches[last_py3_version]
+
+        # Either skip all Py3 versions between 2.7 and the first Py3 version, or set that Py3
+        # version as the lower bound if Py27 not in use.
+        if valid_py27_patches:
+            skipped.extend(f"3.{m}.*" for m in range(0, first_py3_version))
+        else:
+            assert lower_bound is None
+            lower_bound = _determine_lower_bound(3, first_py3_version, first_py3_version_patches[0])
+        assert lower_bound is not None
+
+        # Skip all invalid patches in the first Py3 minor version.
+        skipped.extend(
+            _determine_skipped(
+                3,
+                first_py3_version,
+                first_py3_version_patches,
+                major_minor_is_lower_bound=not valid_py27_patches,
+                major_minor_is_upper_bound=only_one_py3_version,
+            )
+        )
+
+        # If there is only one Py3 version, set that as the upper bound and return.
+        if only_one_py3_version:
+            upper_bound = _determine_upper_bound(
+                3, first_py3_version, first_py3_version_patches[-1]
+            )
+            return _combine_bounds_and_skipped_into_result()
+
+        # Skip any Py3 versions between the lowest Py3 version and last Py3 version.
+        # This `range()` will do nothing if there are no versions between.
+        for py3_minor in range(first_py3_version + 1, last_py3_version):
+            if py3_minor in valid_py3_patches:
+                skipped.extend(
+                    _determine_skipped(
+                        3,
+                        py3_minor,
+                        valid_py3_patches[py3_minor],
+                        major_minor_is_lower_bound=False,
+                        major_minor_is_upper_bound=False,
+                    )
+                )
+            else:
+                skipped.append(f"3.{py3_minor}.*")
+
+        # Finally, set the upper bound to the last Py3 version.
+        upper_bound = _determine_upper_bound(3, last_py3_version, last_py3_version_patches[-1])
+        skipped.extend(
+            _determine_skipped(
+                3,
+                last_py3_version,
+                last_py3_version_patches,
+                major_minor_is_lower_bound=False,
+                major_minor_is_upper_bound=True,
+            )
+        )
+
+        return _combine_bounds_and_skipped_into_result()
+
+    def _determine_all_valid_py2_and_py3_patches(
+        self, interpreter_universe: Iterable[str]
+    ) -> tuple[list[int], dict[int, list[int]]]:
```

Comment:
fyi @chrisjrn , this is what I referred to on https://github.com/pantsbuild/pants/issues/12542